### PR TITLE
Rename dimension

### PR DESCRIPTION
**Pull request by anstaf**
- add `rename_dimesion` to the collection of the SID transformations;
- add the regression test that demonstrate the whole k-axis access to a field.

### CODE REVIEW

**Comment by anstaf** [comment]
launch jenkins

File: tests/regression/whole_axis_access.cpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        using k_pos = in_accessor<2>;
+        using param_list = make_param_list<out, in, k_pos>;
+
+        template <class Eval>
+        GT_FUNCTION static void apply(Eval &&eval) {
+            auto &&res = eval(out());
+            auto k = eval(k_pos());
+            res = 0;
+            for (int kk = 0; kk < k; ++kk)
+                res += eval(in(0, 0, 0, kk));
```
Comment (mbianco) [inline tests/regression/whole_axis_access.cpp:35]: the only thing that bothers me a little is that the dimension to iterate over is the fourth and not the third. If the computation contained more than one stage that have to access the field `in` as regular field, I assume `in` should be passed twice to the computation: once as regular field and once as rename_dimension. Is that right?

Comment (anstaf) [reply tests/regression/whole_axis_access.cpp:35]: Answering the particular question: Yes. If the user needs two semantically different accesses to the data (absolute and relative indexing) he must pass this data twice. This is not that terrible performance wise -- only pointers are duplicated, the data is the same, the caching should work.

Answering generally: You should not expect the perfect solution for such a use case. Just because it is out of stencil computation bounds. However this test demonstrates the known feature of a good API: it is easy to code simple things; corner cases are still possible to implement reliably without backdoors and hacks. Another feature to demonstrate here is that our API is composable -- this solution uses the composition of several features: 1) accessors support offsets for the extra dimensions; 2) computation field interface is well defined by the SID concept; 3) universal (that works on any SID) SID transformations can be implemented with the zero performance cost.

File: include/gridtools/sid/rename_dimension.hpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        namespace rename_dimension_impl_ {
+            template <class OldKey, class NewKey, class Map>
+            auto remap(Map map) {
+                return hymap::convert_to<hymap::keys, meta::replace<get_keys<Map>, OldKey, NewKey>>(std::move(map));
+            }
+
+            template <class OldKey, class NewKey, class Sid>
+            struct renamed_sid : delegate<Sid> {
+                template <class Map>
+                using remaped_t = decltype(remap<OldKey, NewKey>(std::declval<Map>()));
```
Comment (fthaler) [inline include/gridtools/sid/rename_dimension.hpp:31]:
```suggestion
                using remapped_t = decltype(remap<OldKey, NewKey>(std::declval<Map>()));
```

File: include/gridtools/sid/rename_dimension.hpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+                template <class T>
+                renamed_sid(T &&obj) : delegate<Sid>(std::forward<T>(obj)) {}
+
+                friend remaped_t<strides_type<Sid>> sid_get_strides(renamed_sid const &obj) {
```
Comment (fthaler) [inline include/gridtools/sid/rename_dimension.hpp:36]:
```suggestion
                friend remapped_t<strides_type<Sid>> sid_get_strides(renamed_sid const &obj) {
```

File: include/gridtools/sid/rename_dimension.hpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+                    return remap<OldKey, NewKey>(sid_get_strides(obj.impl()));
+                }
+                friend remaped_t<lower_bounds_type<Sid>> sid_get_lower_bounds(renamed_sid const &obj) {
```
Comment (fthaler) [inline include/gridtools/sid/rename_dimension.hpp:39]:
```suggestion
                friend remapped_t<lower_bounds_type<Sid>> sid_get_lower_bounds(renamed_sid const &obj) {
```

File: include/gridtools/sid/rename_dimension.hpp [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+                    return remap<OldKey, NewKey>(sid_get_lower_bounds(obj.impl()));
+                }
+                friend remaped_t<upper_bounds_type<Sid>> sid_get_upper_bounds(renamed_sid const &obj) {
```
Comment (fthaler) [inline include/gridtools/sid/rename_dimension.hpp:42]:
```suggestion
                friend remapped_t<upper_bounds_type<Sid>> sid_get_upper_bounds(renamed_sid const &obj) {
```

File: tests/regression/whole_axis_access.cpp [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+    struct functor {
+        using out = inout_accessor<0>;
+        using in = in_accessor<1, extent<>, 4>;
+        using k_pos = in_accessor<2>;
+        using param_list = make_param_list<out, in, k_pos>;
+
+        template <class Eval>
+        GT_FUNCTION static void apply(Eval &&eval) {
+            auto &&res = eval(out());
```
Comment (fthaler) [inline tests/regression/whole_axis_access.cpp:31]: Performance-wise it would most probably be better to have `res` being a non-reference type and just assign to `out` in the last step (`eval(out()) = res;`). This makes sure that the compiler does compile not generate accesses to main memory for `res` in each loop iteration (it _might_ also be able to optimize them away otherwise as we make use of `restrict`, but no guarantee I guess).

File: tests/regression/whole_axis_access.cpp (hunk 1 above)
Comment (mbianco) [reply tests/regression/whole_axis_access.cpp:35]: Thanks for the answer! I agree with you

**Comment by anstaf** [comment]
launch jenkins
